rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+#### Annotations
+
+Kubernetes annotations will allow users to fine tune transformation of services into Teleport apps. They will override default behaviour, 
+but they are not required for import of services - by default services without any annotations will also be imported.
+
+Annotation `teleport.dev/discovery-type` controls what Teleport entity will be created from this service, for now we will only support value 
+`app` which means Teleport application will be imported from this service. In the future there are plans to expand to database importing.
+If annotation is not set service is treated as an app.
+
+Annotation `teleport.dev/discovery-protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
+default fallback for the the app's URI is `tcp`. However, if annotation is not set, additional attempts will be performed to try to 
+determine protocol of an exposed port:
+-  if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI.
+- if exposed port's name is `http` or it has numeric value 80, `http` will be used.
+- if exposed port's name is `https` or it has numeric value 443, `https` will be used.
+
+Annotation `teleport.dev/discovery-port` controls preferred port for the Kubernetes service, only this one will be used even if service
+has multiple exposed ports. Its value should be one of the exposed service ports; otherwise, the app will not be imported. 
+Value can be matched either by numeric value or by the name of the port defined on the service.
+
+Annotation `teleport.dev/app-rewrite` controls rewrite configuration for Teleport app, if needed. It should
```

Comment:
I don't know if annotations are a good place for this sort of configuration tbh. I think instead of using annotations we need to move all this to a ["discovery config"](https://github.com/gravitational/teleport/pull/26613) resource instead that @marcoandredinis is working on. cc @tigrato

We should also make sure the feature is useful by default and not require users to set annotations or create discovery configs in order for it to work. Those should only be used for fine-tuning.

Reply:
But "discovery config" resource is still about config for discovery service itself, on how to mass process services (which to filter out and which to pick up). And annotations are custom per-service configuration.
And yep, I'm making sure that annotations are not required - I've added it as a first paragraph of the annotations subsection.

Reply:
In this we have to disagree :sweat_smile: 

The problem is that different applications require different rewrite rules and managing that through resource discovery seems a little weird. In the utopian case where each app needs a different config, we will end up having 1 discovery config per app.

The biggest problem is that one of the most installed apps in Kube clusters - grafana - requires custom rewrite configs.

Reply:
Right, okay. I guess Helm keeps the entire resource definitions (for previous versions) in annotations so it shouldn't be that bad.